gh-119993 ignore `NotADirectoryError` in `Path.unlink()` if `missing_ok` is `True` (#120049)
python/cpython: Doc/library/pathlib.rst

   Remove this file or symbolic link.  If the path points to a directory,
   use :func:`Path.rmdir` instead.

>    This method propagates any :exc:`OSError` encountered during removal.

Reviewer: ```suggestion
   If *missing_ok* is false (the default), this method propagates any 
   :exc:`OSError` encountered during removal.
```

Author: Hi @barneygale - is that technically accurate? My understanding is that any `OSError` is propagated regardless of `missing_ok`, which suppresses any `FileNotFoundError`.
I understand the interest to specifically reference the `FileNotFoundError` in the context of `missing_ok` ... how about this new suggestion?

Author: ```
   This method propagates any :exc:`OSError` encountered during removal. 
    For example: If *missing_ok* is false (the default), :exc:`FileNotFoundError` is 
     raised if the path does not exist. 
 
 
     If *missing_ok* is true, :exc:`FileNotFoundError` exceptions will be ...
```

Reviewer: `FileNotFoundError` is a subclass of `OSError`! Maybe we should make that clearer:

```suggestion
   If *missing_ok* is false (the default), this method propagates any 
   :exc:`OSError` from the operating system, including :exc:`FileNotFoundError`.
```

Author: Thanks @barneygale, I've updated with that text and slightly extended the following paragraph:

```rst
   If *missing_ok* is false (the default), this method propagates any 
   :exc:`OSError` from the operating system, including :exc:`FileNotFoundError`.

   If *missing_ok* is true, :exc:`FileNotFoundError` exceptions will be
   ignored (same behavior as the POSIX ``rm -f`` command), any other
   :exc:`OSError` which is encountered will continue to be propogated.
```

   Remove this file or symbolic link.  If the path points to a directory,
   use :func:`Path.rmdir` instead.

   If *missing_ok* is false (the default), this method propagates any

>    :exc:`OSError` from the operating system, including :exc:`FileNotFoundError`.

Reviewer: Let's leave the details for the next paragraph.

```suggestion
   If *missing_ok* is false (the default), this method propagates any
   :exc:`OSError` from the operating system.
```

   If *missing_ok* is true, this shows similar behavior to the POSIX ``rm -f``
   command and any :exc:`FileNotFoundError` or :exc:`NotADirectoryError`
   exceptions will be ignored. This means that the file does not exist after
   execution, but cannot guarantee that the file did exist before. Any other

>    :exc:`OSError` which is encountered will continue to be propogated.

Reviewer: I don't think we need to elaborate so much!

```suggestion
   If *missing_ok* is true, :exc:`FileNotFoundError` and 
   :exc:`NotADirectoryError` exceptions will be ignored. This behavior is
   similar to the POSIX ``rm -f`` command.
```

   .. versionchanged:: 3.8
      The *missing_ok* parameter was added.

   .. versionchanged:: 3.??
      The *missing_ok* parameter will also ignore :exc:`NotADirectoryError`

> 

Reviewer: ```suggestion
   .. versionchanged:: 3.14
      Suppresses :exc:`NotADirectoryError` exceptions when *missing_ok* is
      true.
```
